cdap-authorization-ldap-role/src/main/java/io/cdap/cdap/security/authorization/ldap/role/searcher/LDAPSearcher.java

```diff
@@ -0,0 +1,204 @@
+    this.config = config;
+    properties = getConnectionProperties();
+    baseDNList = config.getSearchBaseDn().split(LDAPConstants.BASE_DN_SPLITTER);
+  }
+
+  /**
+   * Tests connection to LDAP
+   */
+  public void testConnection() {
+    try {
+      DirContext context = getConnection();
+      context.close();
+    } catch (NamingException e) {
+      String errorMsg = String.format("Failed to establish connection to '%s'", config.getUrl());
+      throw new RuntimeException(errorMsg, e);
+    }
+  }
+
+  /**
+   * Searches for groups by username
+   *
+   * @param username Name of user
+   * @return Set of groups
+   */
+  public Set<String> searchGroups(String username) {
+    Set<String> groups = Collections.emptySet();
+
+    boolean retry = true;
+    NamingException exception = null;
+    while (retry) {
+      try {
+        DirContext context = getConnection();
+
+        SearchControls controls = new SearchControls();
+        if (config.isRecursiveSearch()) {
+          controls.setSearchScope(SUBTREE_SCOPE);
+        }
+
+        groups = Arrays.stream(baseDNList)
+          .map(baseDN -> searchGroups(baseDN, username, context, controls))
+          .flatMap(Collection::stream)
+          .collect(Collectors.toSet());
+
+        retry = false;
+        context.close();
+      } catch (NamingException e) {
+        String errorMsg = String.format("Failed to find groups for '%s': %s", username, e.getCause());
+        if (Objects.nonNull(exception)) {
+          throw new RuntimeException(errorMsg, e);
+        }
+
+        exception = e;
+        LOG.warn(errorMsg);
+      }
+    }
+
+    return groups;
+  }
+
+  private Set<String> searchGroups(String baseDN, String username, DirContext context, SearchControls controls) {
+    String filter = String.format(config.getSearchFilter(), username);
+    Set<String> groups = new HashSet<>();
+
+    try {
+      NamingEnumeration<SearchResult> renum = context.search(baseDN, filter, controls);
+
+      if (!renum.hasMore()) {
+        LOG.debug("Cannot locate user information for '{}'", username);
+        return groups;
+      }
+
+      SearchResult result = renum.next();
+
+      Attribute memberOf = result.getAttributes().get(config.getMemberAttribute());
+      if (memberOf != null) {
+        for (int i = 0; i < memberOf.size(); i++) {
+          groups.add(memberOf.get(i).toString());
+        }
+      }
+    } catch (PartialResultException e) {
+      LOG.warn("Failed to find groups for '{}' in '{}'", username, baseDN);
+    } catch (NamingException e) {
+      String errorMsg = String.format("Failed to find groups for '%s' in '%s'", username, baseDN);
+      throw new RuntimeException(errorMsg, e);
+    }
+
+    return groups;
+  }
+
+  private DirContext getConnection() throws NamingException {
+    NamingException exception = null;
+    for (int i = 1; i <= LDAPConstants.MAX_RETRIES; i++) {
+      try {
+        return new InitialDirContext(properties);
```

Comment:
does the pool still throw an exception here?

Reply:
Yes, for example if host is not reachable.